File: learning-akka/akkademy-server/src/main/java/book/java/akkademy/AkkademyDb.java
package book.java.akkademy;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import book.kotlin.messages.*;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

import java.util.HashMap;
import java.util.Map;

public class AkkademyDb extends AbstractActor {

	final LoggingAdapter log = Logging.getLogger(context().system(), this);
	final Map<String, Object> map = new HashMap<>();

	@Override
	public PartialFunction<Object, BoxedUnit> receive() {
		return ReceiveBuilder.
				match(SetRequest.class, message -> {
					log.info("Received set request – key: {}; value: {}", message.getKey(), message.getValue());
					map.put(message.getKey(), message.getValue());
					sender().tell(new Status.Success(message.getKey()), self());
				}).
				match(GetRequest.class, message -> {
					log.info("Received get request – key: {}", message.getKey());
					Object value = map.get(message.getKey());
					Object response = (value != null)
							? value
							: new Status.Failure(new KeyNotFoundException(message.getKey()));
					sender().tell(response, self());
				}).
				match(SetIfNotExists.class, message -> {
					log.info("Received SetIfNotExists request – key: {}; value: {}", message.getKey(), message.getValue());
					Object previousValue = map.putIfAbsent(message.getKey(), message.getValue());
					Object response = previousValue == null
							? new Status.Success(message.getKey())
							: new Status.Failure(new KeyAlreadyExistsException(message.getKey()));
					sender().tell(response, self());
				}).
				match(DeleteRequest.class, message -> {
					log.info("Received DeleteRequest – key: {}", message.getKey());
					Object removed = map.remove(message.getKey());
					Object response = removed != null
							? new Status.Success(removed)
							: new Status.Failure(new KeyNotFoundException(message.getKey()));
					sender().tell(response, self());
				}).
				matchAny(o -> {
					log.info("Received unknown message: {}", o);
					sender().tell(new Status.Failure(new ClassNotFoundException(o.getClass().toString())), self());
				})
				.build();
	}

}

File: learning-akka/akkademy-server/src/main/kotlin/book/kotlin/akkademy/AkkademyDb.kt
package book.kotlin.akkademy

import akka.actor.Status
import akka.event.Logging
import book.kotlin.messages.*
import book.kotlin.util.KotlinActor
import book.kotlin.util.system

class AkkademyDb : KotlinActor() {

	val log = Logging.getLogger(context.system, this)
	val map = mutableMapOf<String, Any>()

	override fun onReceive(message: Any) = when (message) {
		is SetRequest -> {
			log.info("received SetRequest - key: {}; value: {}", message.key, message.value)
			map += message.key to message.value
			sender.tell(Status.Success(message.key))
		}
		is GetRequest -> {
			log.info("Received get request – key: {}", message.key)
			val value = map[message.key]
			val response = value ?: Status.Failure(KeyNotFoundException(message.key))
			sender.tell(response)
		}
		is SetIfNotExists -> {
			log.info("Received SetIfNotExists request – key: {}; value: {}", message.key, message.value)
			val previousValue = map.putIfAbsent(message.key, message.value)
			val response: Status.Status =
					if (previousValue == null) Status.Success(message.key)
					else Status.Failure(KeyAlreadyExistsException(message.key))
			sender.tell(response)
		}
		is DeleteRequest -> {
			log.info("Received DeleteRequest – key: {}", message.key)
			val removed = map.remove(message.key)
			val response: Status.Status =
					if (removed != null) Status.Success(removed)
					else Status.Failure(KeyNotFoundException(message.key))
			sender.tell(response)
		}
		else -> {
			log.info("received unknown message: {}", message)
			sender.tell(Status.Failure(ClassNotFoundException(message::class.toString())))
		}
	}

}
